File: 3093-longest-common-suffix-queries/3093-longest-common-suffix-queries.ts
class TrieNode {
    children: Map<string, TrieNode>;
    bestIndex: number;
    bestLength: number;

    constructor() {
        this.children = new Map();
        this.bestIndex = -1;
        this.bestLength = Infinity;
    }
}

// 두 인덱스의 우선순위를 비교하는 헬퍼 함수 (길이가 짧은 것 우선, 같으면 인덱스가 작은 것 우선)
function isBetter(
    newLen: number, newIdx: number,
    oldLen: number, oldIdx: number
): boolean {
    if (newLen !== oldLen) {
        return newLen < oldLen;
    }
    return newIdx < oldIdx;
}

function stringIndices(wordsContainer: string[], wordsQuery: string[]): number[] {
    const root = new TrieNode();

    // 1. 전체 container에서 기본 fallback으로 사용할 가장 최적의 인덱스 찾기
    let bestGlobalIdx = 0;
    for (let i = 1; i < wordsContainer.length; i++) {
        if (isBetter(wordsContainer[i].length, i, wordsContainer[bestGlobalIdx].length, bestGlobalIdx)) {
            bestGlobalIdx = i;
        }
    }

    // 2. 단어들을 뒤집어서 Trie에 삽입
    for (let i = 0; i < wordsContainer.length; i++) {
        const word = wordsContainer[i];
        const len = word.length;
        let node = root;

        // 루트 노드 업데이트 (공통 접미사가 전혀 없는 경우의 대비책)
        if (isBetter(len, i, node.bestLength, node.bestIndex)) {
            node.bestIndex = i;
            node.bestLength = len;
        }

        // 뒤에서부터 문자 삽입 (접미사 매칭을 위해)
        for (let j = word.length - 1; j >= 0; j--) {
            const char = word[j];
            if (!node.children.has(char)) {
                node.children.set(char, new TrieNode());
            }
            node = node.children.get(char)!;

            // 해당 노드를 거쳐가는 단어 중 가장 최적의 값 저장
            if (isBetter(len, i, node.bestLength, node.bestIndex)) {
                node.bestIndex = i;
                node.bestLength = len;
            }
        }
    }

    // 3. 쿼리 처리
    const result: number[] = [];
    for (const query of wordsQuery) {
        let node = root;
        let lastValidIndex = bestGlobalIdx;

        // 쿼리 문자열을 뒤에서부터 탐색
        for (let i = query.length - 1; i >= 0; i--) {
            const char = query[i];
            if (!node.children.has(char)) {
                break;
            }
            node = node.children.get(char)!;
            lastValidIndex = node.bestIndex;
        }
        result.push(lastValidIndex);
    }

    return result;
}

File: 3093-longest-common-suffix-queries/3093-longest-common-suffix-queries.java
class Solution {
    // Trie 노드 정의
    static class TrieNode {
        TrieNode[] children = new TrieNode[26];
        int bestIndex = -1;
        int bestLength = Integer.MAX_VALUE;
    }

    public int[] stringIndices(String[] wordsContainer, String[] wordsQuery) {
        TrieNode root = new TrieNode();
        
        // 1. 공통 접미사가 전혀 없을 때 사용할 글로벌 기본값 찾기
        int bestGlobalIdx = 0;
        for (int i = 1; i < wordsContainer.length; i++) {
            if (isBetter(wordsContainer[i].length(), i, wordsContainer[bestGlobalIdx].length(), bestGlobalIdx)) {
                bestGlobalIdx = i;
            }
        }

        // 2. wordsContainer의 단어들을 뒤집어서 Trie에 삽입
        for (int i = 0; i < wordsContainer.length; i++) {
            String word = wordsContainer[i];
            int len = word.length();
            TrieNode node = root;

            // 루트 노드 조건 업데이트
            if (isBetter(len, i, node.bestLength, node.bestIndex)) {
                node.bestIndex = i;
                node.bestLength = len;
            }

            // 단어를 뒤에서부터 탐색하여 Trie에 삽입
            for (int j = len - 1; j >= 0; j--) {
                int charIdx = word.charAt(j) - 'a';
                if (node.children[charIdx] == null) {
                    node.children[charIdx] = new TrieNode();
                }
                node = node.children[charIdx];

                // 현재 노드를 공유하는 가장 최적의 단어 정보 갱신
                if (isBetter(len, i, node.bestLength, node.bestIndex)) {
                    node.bestIndex = i;
                    node.bestLength = len;
                }
            }
        }

        // 3. wordsQuery 처리
        int[] result = new int[wordsQuery.length];
        for (int i = 0; i < wordsQuery.length; i++) {
            String query = wordsQuery[i];
            TrieNode node = root;
            int lastValidIndex = bestGlobalIdx;

            // 쿼리 단어를 뒤에서부터 탐색 (괄호 추가 완료)
            for (int j = query.length() - 1; j >= 0; j--) {
                int charIdx = query.charAt(j) - 'a';
                if (node.children[charIdx] == null) {
                    break;
                }
                node = node.children[charIdx];
                lastValidIndex = node.bestIndex;
            }
            result[i] = lastValidIndex;
        }

        return result;
    }

    // 우선순위 비교 헬퍼 메소드 (길이가 짧은 것 우선, 같으면 인덱스가 작은 것 우선)
    private boolean isBetter(int newLen, int newIdx, int oldLen, int oldIdx) {
        if (newLen != oldLen) {
            return newLen < oldLen;
        }
        return newIdx < oldIdx;
    }
}
